### ISSUE TITLE
Operator parse error should be scope error

### ISSUE THREAD

**Issue by andreasabel** [agda/agda#2591]
To parse operator expressions, we consider exactly the operators which are in scope.  For the user, this can be confusing.
```agda
mutual

  infixl 6 _+_

  data D : Set where
    _+_ : (x y : D) → D
    bla : ∀ x y z → E x y z  → D

  postulate
    f : ∀ x y z →  _+_ x y  ≡ z

  data E (x y : D) : D → Set where
    e : ∀ z → (f : _+_ x y ≡ z) → E x y z
```
This gives a parse error in the postulate.
```
Could not parse the application _+_ x y ≡ z
Operators used in the grammar:
  ≡ (infix operator, level 4)
```
Agda pretends here it does not know anything about the operator `_+_`, even though it has seen the infix declaration, and the constructor declaration is above the postulate.

It would be better if parsing succeeded and a scope error would be raised subsequently.

We currently conflate the set of known operators with the set of operators in scope, but this identification is not mandatory.  We could keep a separate set of known operators.

**Comment by nad** [comment]
What do you suggest should happen if another instance of `_+_` is in scope? Should both operators be part of the generated grammar? This might lead the parser to incorrectly conclude that the code is ambiguous.

Perhaps one could parse the expression again, and include the "known" operators, if the first run of the parser does not yield any valid expression. If the second run yields exactly one (or at least one?) valid expression, then one could complain about any operators that are not in scope: `parse error; note that _+_ and _*_ are not in scope`.

**Comment by andreasabel** [comment]
We could get rid of name ambiguities during scope checking, once we have parsed.

Maybe the more worrying problem is that when parsing x + y * z with _+_ and _*_ in scope and a third operator _+_*_ in the set of known operators, parsing would lead to ambiguous syntax trees.  If we return all of the possible parses, we might get to a combinatorial explosion.

Your suggestion to reparse when parsing fails seems to make most sense.

**Comment by UlfNorell** [comment]
It's indeed a bit strange that you get `Couldn't parse the application` in your case but
```agda
    f : ∀ x y z e → bla x y z e ≡ z
```
gives you `Not in scope: bla`.

It's not clear, however, how you would define this set of "known" operators.
- Private operators from local (imported?) modules?
- Public operators from local (imported?) modules that are not opened?
- From modules that are opened but hiding/not using/renaming the operator?
- Operators defined further down in the module, but whose fixity declaration appears above?
- Operators from transitively imported modules?
- Operators defined in a paper that you cite in a comment? *(probably not this one)*

**Comment by UlfNorell** [comment]
We really should just give an out of scope error for `_+_` here, since it's clear that `_+_` is not part of any operator in scope.

**Comment by nad** [comment]
Is the idea to check, for every name part in an expression, whether it is in scope (possibly using a module prefix, if applicable) as a regular name, as part of an operator, or as part of some `syntax` notation? A naive implementation might be rather slow if lots of operators are in scope. Even if we construct a reasonably efficient implementation I suggest that we optimise for the common case and only run this algorithm if operator parsing fails.

**Comment by UlfNorell** [comment]
Clearly we already do this check:
```agda
postulate
  _bla_ : Set → Set → Set
  foo : bla Set Set -- Could not parse application
```
```agda
postulate
  _blap_ : Set → Set → Set
  foo : bla Set Set -- Not in scope: bla
```

**Comment by nad** [comment]
No, we don't verify that, say, `]_` is "in scope" given that `M._[_]_` is in scope.

**Comment by UlfNorell** [comment]
But we do it for non-qualified operators, as evidenced by my example.

All I'm suggesting is that
```agda
postulate foo : NOTINSCOPE Set
```
and
```agda
postulate foo : NOT_IN_SCOPE Set
```
both give a `Not in scope` error. I don't see how the latter case would require expensive work, but the former not.

**Comment by nad** [comment]
What about the following code?
```agda
postulate foo : ACTUALLY_IN Set SCOPE
```

**Comment by UlfNorell** [comment]
That should behave the same as `ACTUALLYIN Set SCOPE`, I don't see why the underscore should make a difference. In this case I suppose it would be nice if it complained about both `ACTUALLY_IN` and `SCOPE` not being in scope.

**Comment by nad** [comment]
If `ACTUALLY_IN_SCOPE` is in scope, then we should not complain about the code I gave above.

**Comment by UlfNorell** [comment]
No, of course not. We also shouldn't complain about `x + y` if `_+_`, `x` and `y` are all in scope. I fail to see your point.

**Comment by nad** [comment]
All I'm saying is that we can't just look up every name part in the symbol table, and a linear search might be too slow.

**Comment by UlfNorell** [comment]
Ok, but then I don't understand how these two examples can behave differently:
```agda
postulate
  _bla_ : Set → Set → Set
  foo : bla Set Set -- Could not parse application
```
```agda
postulate
  _blap_ : Set → Set → Set
  foo : bla Set Set -- Not in scope: bla
```

**Comment by nad** [comment]
In the former case `bla` cannot be parsed as an atom, but in the latter case it can:
https://github.com/agda/agda/blob/c727f3f19b0418e218cdf70ebff66f822c01f415/src/full/Agda/Syntax/Concrete/Operators.hs#L258-L288

**Comment by UlfNorell** [comment]
former <-> latter?

Why is this distinction cheap for `bla` and expensive for `_+_`?

**Comment by nad** [comment]
> former <-> latter?

No. In the latter case we parse the application of `bla` to `Set` and `Set`, and then we notice that `bla` is not in scope. In the former case we fail to parse anything at all.

> Why is this distinction cheap for `bla` and expensive for `_+_`?

I don't understand the question.

**Comment by UlfNorell** [comment]
My question is: If we can parse `bla` as an atom and then notice it's not in scope, why can't we do the same for `_+_`?
```agda
postulate
  _ : bla Set -- Not in scope: bla
  _ : _+_ Set -- Could not parse the application
```

**Comment by nad** [comment]
Perhaps we can. However, when expressions are parsed we currently do not treat `_+_` as a single token, but as the three tokens `_` (Left), `+` and `_` (Right). When adding support for sections I used this technique to avoid creating grammars of size Ω(2<sup>*n*</sup>), where *n* is the number of internal holes of the operator/notation with the largest number of holes.

**Comment by andreasabel** [comment]
Seems like the OP is fixed on master (2.9.0), I am getting the good error:
```
error: [NotInScope]
Not in scope:
  _+_ at <the postulate>
when scope checking _+_
```
It is still broken in 2.8.0.
We should try to learn what caused this behavioral change (for the good).

I am running a bisect now.

**Comment by andreasabel** [comment]
Bisection attributes this to commit 21154b1e4eb9e67115c7175a81ba4bcff9a820e2 by @AndrasKovacs .
> Date:   Tue Sep 9 19:34:47 2025 +0200
>
>    Fix operator parsing asymptotics

- PR #8079

The PR contains a similar change to the test case of 
- #1977 

see https://github.com/agda/agda/issues/1977#issuecomment-3271029986.  I acknowledged this change in https://github.com/agda/agda/pull/8079/changes#r2333864143 (but this was 5 month ago so I didn't remember now).

I will add the OP as testcase.
